### PR TITLE
Issue #62 fix query params

### PR DESCRIPTION
**Pull request by istaveren**
Fix for issue #62, #63 & #64 

Also added some tests.

### CODE REVIEW

**Comment by istaveren** [comment]
I checked with 
php-cs-fixer fix tests/ --rules=@PSR2
Runnig it with .php_cs failed with
PHP Fatal error:  Uncaught Error: Class 'Symfony\CS\Config\Config' not found in /home/iwan/projects/prive/pdo-via-oci8/.php_cs:73

File: tests/ConnectionTest.php [hunk 1]
```diff
@@ -5,14 +5,181 @@
...
+    protected $con = null;
+
+    /**
+     * Set up a new object
+     *
+     * @return null
+     */
+    public function setUp()
+    {
+        $user = getenv("OCI_USER") ?: 'oracle';
```
Comment (yajra) [inline tests/ConnectionTest.php]: Username is `system` while password is `oracle`.

Comment (istaveren) [reply tests/ConnectionTest.php]: Changed

**Comment by yajra** [comment]
Can you please fix the conflicts and update the PR from 1.0 branch? I already fixed the docker used by the package for it to work with TravisCI. Thanks!

**Comment by istaveren** [comment]
Conflicts fixed.

**Comment by yajra** [comment]
Just tested this on my project and it seems like the patch for bindParam is failing the app.

```
ErrorException::("oci_bind_by_name(): ORA-01036: illegal variable name/number")

oci_bind_by_name(":p-1", "platform")
```

**Comment by istaveren** [comment]
Can you write a test or add an example on how I can reproduce it?

**Comment by yajra** [comment]
I am using this package along with my [other one](https://github.com/yajra/laravel-oci8) on Laravel apps. I think you can recreate the issue by creating a fresh laravel app then just run the migration.

**Comment by istaveren** [comment]
Yes, I can probably do that. But would it not be easy to just write a failing unit test? 
So that I can fix it?

**Comment by yajra** [comment]
Oh yeah, writing a failing test is a great idea. Will try to create one.  Btw, you need some seeder for the laravel side for some basic inserts. Thanks!

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by istaveren** [comment]
I would love to fix this, can we chat? Please ping me on hangout istaveren or some other channel.

**Comment by yajra** [comment]
Can't reproduce a failing tests but it seems like there are some modifications on my other package that is need to work with this. Creating a PR for it and we'll see how it works.

**Comment by yajra** [comment]
Or actually, this is how it fails:

```php
    public function testBindParamName()
    {
        $stmt = $this->con->prepare('select * from all_tables where upper(owner) = upper(?) and upper(table_name) = upper(?)');
        $owner = 'owner';
        $table = 'table';
        $this->assertTrue($stmt->bindParam(0, $owner, PDO::PARAM_STR));
        $this->assertTrue($stmt->bindParam(1, $table, PDO::PARAM_STR));
    }
```

But as you've stated, the param should start at `1`?

**Comment by yajra** [comment]
PR https://github.com/yajra/laravel-oci8/pull/505 sent but it's failing on RefCursor tests. Will try to check again further when I got the chance.

**Comment by istaveren** [comment]
Correct see issue #63 a bindParam should start with 1.

https://www.php.net/manual/en/pdostatement.bindparam.php

File: src/Pdo/Oci8/Statement.php [hunk 2]
```diff
@@ -417,7 +418,7 @@ public function bindParam($parameter, &$variable, $dataType = PDO::PARAM_STR, $m
 
         // Replace the first @oci8param to a pseudo named parameter
         if (is_numeric($parameter)) {
-            $parameter = ':p' . $parameter;
+            $parameter = ':p' . ($parameter-1);
```
Comment (yajra) [inline src/Pdo/Oci8/Statement.php]: Ok, I think this should be reverted since we need to use a colon-prefixed bind variable placeholder `bv_name` as per https://php.net/manual/en/function.oci-bind-by-name.php. 

See the return of bindParam for ref:

```php
return oci_bind_by_name($this->sth, $parameter, $variable, $maxLength, $ociType);
```

Given this, I think bindParam using index is **not yet supported** and should be address by another PR.

**Comment by yajra** [comment]
Or I might be missing something that we can adjust to support bindParam using int?

File: src/Pdo/Oci8/Statement.php (hunk 2 above)
Comment (istaveren) [reply src/Pdo/Oci8/Statement.php]: Yes but that works.
The problem is that you try to send param 0 which is invalid.

Comment (yajra) [reply src/Pdo/Oci8/Statement.php]: Actually, it would still work with either 0 or 1 as base index since **we alway convert numeric keys to named parameters**. It just that it is not the standard way and can be adjusted on my other package. Already sent a PR related to this on https://github.com/yajra/pdo-via-oci8/pull/66 and https://github.com/yajra/laravel-oci8/pull/505.

Anyways, I suggest this line to be reverted unless we find a solution for it to work?

**Comment by istaveren** [comment]
So how can this be fixed?

Both param 0 and Param 1 should point to the first param?
I think in that case your code will still not work, because your code assumes 1 to be the second one where it is the first one. By the PDO definition. 

> Or actually, this is how it fails:
> 
> ```
>     public function testBindParamName()
>     {
>         $stmt = $this->con->prepare('select * from all_tables where upper(owner) = upper(?) and upper(table_name) = upper(?)');
>         $owner = 'owner';
>         $table = 'table';
>         $this->assertTrue($stmt->bindParam(0, $owner, PDO::PARAM_STR));
>         $this->assertTrue($stmt->bindParam(1, $table, PDO::PARAM_STR));
>     }
> ```
> 
> But as you've stated, the param should start at `1`?

**Comment by yajra** [comment]
Sorry I've provided a bad example. Basically, this kind of syntax is not yet supported and thus should be removed in this PR.

```php
$stmt = $this->con->prepare(
  'select * from all_tables where upper(owner) = upper(?) and upper(table_name) = upper(?)'
);

// this example is bad.
$this->assertTrue($stmt->bindParam(0, $owner, PDO::PARAM_STR));
$this->assertTrue($stmt->bindParam(1, $table, PDO::PARAM_STR));

// this is how it would be processed now.
$stmt = $this->con->prepare(
  'select * from all_tables where upper(owner) = upper(?) and upper(table_name) = upper(?)'
);

// after prepare, sql becomes
'select * from all_tables where upper(owner) = upper(:p0) and upper(table_name) = upper(:p1)'

$this->assertTrue($stmt->bindParam(":p0", $owner, PDO::PARAM_STR));
$this->assertTrue($stmt->bindParam(":p1", $table, PDO::PARAM_STR));
```

**Comment by istaveren** [comment]
Ok reverted it, will create a pull request on #64

**Comment by yajra** [comment]
Released on v1.4.0, thanks a lot! 🍻
